SM2135 support for smart led

<!-- READ THIS FIRST:
 - This is for feature requests only, for issues please go to the issues repository.
 - Please be as descriptive as possible, especially use-cases that can otherwise not be solved boost the problem's priority.

 DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Describe the problem you have/What new integration you would like**
I bought some LSC GU10 Smart LED from Action and would like to use ithem with ESPHome. I've successfully used Tuya-convert but it seems that support for SM2135 is missing from ESPHome

**Please describe your use case for this integration and alternatives you've tried:**
It seems that some smart bulb are using this chip. Would be nice to use ESPHome on it.

**Additional context**
See this issue at Tasmota: https://github.com/arendst/Sonoff-Tasmota/issues/6495
A lot of information can be found there

<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE FEATURE REQUEST WILL BE CLOSED AUTOMATICALLY. -->

@RoelVideler 

I tried to upload to compile but I get this error

`Archiving /data/kantoor1/.pioenvs/kantoor1/libFrameworkArduino.a
Linking /data/kantoor1/.pioenvs/kantoor1/firmware.elf
/data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function `esphome::sm2135::SM2135::dump_config()':
sm2135.cpp:(.text._ZN7esphome6sm21356SM213511dump_configEv+0x44): multiple definition of `esphome::sm2135::SM2135::dump_config()'
/data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM213511dump_configEv+0x44): first defined here
/data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function `esphome::sm2135::SM2135::loop()':
sm2135.cpp:(.text._ZN7esphome6sm21356SM21354loopEv+0x18): multiple definition of `esphome::sm2135::SM2135::loop()'
/data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM21354loopEv+0x18): first defined here
/data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function `esphome::sm2135::SM2135::setup()':
sm2135.cpp:(.text._ZN7esphome6sm21356SM21355setupEv+0x10): multiple definition of `esphome::sm2135::SM2135::setup()'
/data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM21355setupEv+0x10): first defined here
collect2: error: ld returned 1 exit status
*** [/data/kantoor1/.pioenvs/kantoor1/firmware.elf] Error 1
========================= [FAILED] Took 54.95 seconds =========================`

Am I doing something wrong?

`esphome:
  name: kantoor1
  platform: ESP8266
  board: esp01_1m
  includes:
    - custom_components/sm2135/sm2135.h
    - custom_components/sm2135/sm2135.cpp

sm2135:
  data_pin: GPIO12
  clock_pin: GPIO14
  
wifi:
  ssid: "blurredout"
  password: "temp123"
  
  manual_ip:
    static_ip: 192.168.50.85
    gateway: 192.168.50.1
    subnet: 255.255.255.0
    dns1: 192.168.50.1

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Tuya Fallback Hotspot"
    password: "ObPG9RIlmLIg"

binary_sensor:
  - platform: status
    name: "kantoor1_status"

captive_portal:

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:

# Individual outputs
output:
  - platform: sm2135
    id: output_red
    channel: 0
  - platform: sm2135
    id: output_green
    channel: 1
  - platform: sm2135
    id: output_blue
    channel: 2
  - platform: sm2135
    id: output_white
    channel: 3
  - platform: sm2135
    id: output_warmwhite
    channel: 4

light:
  - platform: rgbww
    name: "kantoor1"
    default_transition_length: 1s
    cold_white: output_white
    warm_white: output_warmwhite
    cold_white_color_temperature: 6536 K
    warm_white_color_temperature: 2000 K
    restore_mode: ALWAYS_ON
    red: output_red
    green: output_green
    blue: output_blue`

@RoelVideler 

I tried it without the include line and it gives this error.

`/data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function `esphome::sm2135::SM2135::dump_config()':
sm2135.cpp:(.text._ZN7esphome6sm21356SM213511dump_configEv+0x44): multiple definition of `esphome::sm2135::SM2135::dump_config()'
/data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM213511dump_configEv+0x44): first defined here
/data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function `esphome::sm2135::SM2135::loop()':
sm2135.cpp:(.text._ZN7esphome6sm21356SM21354loopEv+0x18): multiple definition of `esphome::sm2135::SM2135::loop()'
/data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM21354loopEv+0x18): first defined here
/data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function `esphome::sm2135::SM2135::setup()':
sm2135.cpp:(.text._ZN7esphome6sm21356SM21355setupEv+0x10): multiple definition of `esphome::sm2135::SM2135::setup()'
/data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM21355setupEv+0x10): first defined here
collect2: error: ld returned 1 exit status
*** [/data/kantoor1/.pioenvs/kantoor1/firmware.elf] Error 1
========================== [FAILED] Took 2.74 seconds ==========================`

But I wanted to try your cpp file cause of the bug fixes you described

I’m still seeing multiple locations of the cpp file, make sure you only
have one copy, same goes for the other files. You could even remove the new
one and see if it compiles, if it does you have another sm2135 driver
installed somewhere. Let me know if that doesn’t work.

On Tue, 24 Mar 2020 at 18:22, hiddedb <notifications@github.com> wrote:

> @RoelVideler <https://github.com/RoelVideler>
>
> I tried it without the include line and it gives this error.
>
> /data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In function
> esphome::sm2135::SM2135::dump_config()':
> sm2135.cpp:(.text._ZN7esphome6sm21356SM213511dump_configEv+0x44): multiple
> definition of esphome::sm2135::SM2135::dump_config()'
> /data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM213511dump_configEv+0x44):
> first defined here /data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In
> function esphome::sm2135::SM2135::loop()':
> sm2135.cpp:(.text._ZN7esphome6sm21356SM21354loopEv+0x18): multiple
> definition of esphome::sm2135::SM2135::loop()'
> /data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM21354loopEv+0x18):
> first defined here /data/kantoor1/.pioenvs/kantoor1/src/sm2135.cpp.o: In
> function esphome::sm2135::SM2135::setup()':
> sm2135.cpp:(.text._ZN7esphome6sm21356SM21355setupEv+0x10): multiple
> definition of esphome::sm2135::SM2135::setup()'
> /data/kantoor1/.pioenvs/kantoor1/src/esphome/components/sm2135/sm2135.cpp.o:sm2135.cpp:(.text._ZN7esphome6sm21356SM21355setupEv+0x10):
> first defined here collect2: error: ld returned 1 exit status ***
> [/data/kantoor1/.pioenvs/kantoor1/firmware.elf] Error 1
> ========================== [FAILED] Took 2.74 seconds
> ==========================
>
> But I wanted to try your cpp file cause of the bug fixes you described
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esphome/feature-requests/issues/443#issuecomment-603385765>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGJAHUVN7LNJUH3VNXOFUVTRJDT5XANCNFSM4JC45C5Q>
> .
>

--- a/CODEOWNERS
+++ b/CODEOWNERS
@@ -85,6 +85,7 @@ esphome/components/sgp40/* @SenexCrenshaw
 esphome/components/sht4x/* @sjtrny
 esphome/components/shutdown/* @esphome/core
 esphome/components/sim800l/* @glmnet
+esphome/components/sm2135/* @BoukeHaarsma23
 esphome/components/spi/* @esphome/core
 esphome/components/ssd1322_base/* @kbx81
 esphome/components/ssd1322_spi/* @kbx81

--- a/esphome/components/sm2135/__init__.py
+++ b/esphome/components/sm2135/__init__.py
@@ -1,0 +1,31 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import pins
+from esphome.const import (
+    CONF_CLOCK_PIN,
+    CONF_DATA_PIN,
+    CONF_ID,
+)
+
+AUTO_LOAD = ["output"]
+sm2135_ns = cg.esphome_ns.namespace("sm2135")
+SM2135 = sm2135_ns.class_("SM2135", cg.Component)
+
+MULTI_CONF = True
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(SM2135),
+        cv.Required(CONF_DATA_PIN): pins.gpio_output_pin_schema,
+        cv.Required(CONF_CLOCK_PIN): pins.gpio_output_pin_schema,
+    }
+).extend(cv.COMPONENT_SCHEMA)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+
+    data = yield cg.gpio_pin_expression(config[CONF_DATA_PIN])
+    cg.add(var.set_data_pin(data))
+    clock = yield cg.gpio_pin_expression(config[CONF_CLOCK_PIN])
+    cg.add(var.set_clock_pin(clock))

--- a/esphome/components/sm2135/__init__.py
+++ b/esphome/components/sm2135/__init__.py
@@ -8,6 +8,8 @@ from esphome.const import (
 )
 
 AUTO_LOAD = ["output"]
+CODEOWNERS = ["@BoukeHaarsma23"]
+
 sm2135_ns = cg.esphome_ns.namespace("sm2135")
 SM2135 = sm2135_ns.class_("SM2135", cg.Component)
 

--- a/esphome/components/sm2135/output.py
+++ b/esphome/components/sm2135/output.py
@@ -1,0 +1,27 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import output
+from esphome.const import CONF_CHANNEL, CONF_ID
+from . import SM2135
+
+DEPENDENCIES = ["sm2135"]
+
+Channel = SM2135.class_("Channel", output.FloatOutput)
+
+CONF_SM2135_ID = "sm2135_id"
+CONFIG_SCHEMA = output.FLOAT_OUTPUT_SCHEMA.extend(
+    {
+        cv.GenerateID(CONF_SM2135_ID): cv.use_id(SM2135),
+        cv.Required(CONF_ID): cv.declare_id(Channel),
+        cv.Required(CONF_CHANNEL): cv.int_range(min=0, max=65535),
+    }
+).extend(cv.COMPONENT_SCHEMA)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield output.register_output(var, config)
+
+    parent = yield cg.get_variable(config[CONF_SM2135_ID])
+    cg.add(var.set_parent(parent))
+    cg.add(var.set_channel(config[CONF_CHANNEL]))

--- a/esphome/components/sm2135/output.py
+++ b/esphome/components/sm2135/output.py
@@ -5,6 +5,7 @@ from esphome.const import CONF_CHANNEL, CONF_ID
 from . import SM2135
 
 DEPENDENCIES = ["sm2135"]
+CODEOWNERS = ["@BoukeHaarsma23"]
 
 Channel = SM2135.class_("Channel", output.FloatOutput)
 

--- a/esphome/components/sm2135/sm2135.cpp
+++ b/esphome/components/sm2135/sm2135.cpp
@@ -1,0 +1,84 @@
+#include "sm2135.h"
+#include "esphome/core/log.h"
+
+
+// Tnx to the work of https://github.com/arendst (Tasmota) for making the initial version of the driver
+
+
+namespace esphome {
+namespace sm2135 {
+
+static const char *TAG = "sm2135";
+
+static const uint8_t SM2135_ADDR_MC   =  0xC0;  // Max current register
+static const uint8_t SM2135_ADDR_CH   =  0xC1;  // RGB or CW channel select register
+static const uint8_t SM2135_ADDR_R    =  0xC2;  // Red color
+static const uint8_t SM2135_ADDR_G    =  0xC3;  // Green color
+static const uint8_t SM2135_ADDR_B    =  0xC4;  // Blue color
+static const uint8_t SM2135_ADDR_C    =  0xC5;  // Cold
+static const uint8_t SM2135_ADDR_W    =  0xC6;  // Warm
+
+static const uint8_t SM2135_RGB       =  0x00;  // RGB channel
+static const uint8_t SM2135_CW        =  0x80;  // CW channel (Chip default)
+
+static const uint8_t SM2135_10MA      =  0x00;
+static const uint8_t SM2135_15MA      =  0x01;
+static const uint8_t SM2135_20MA      =  0x02;  // RGB max current (Chip default)
+static const uint8_t SM2135_25MA      =  0x03;
+static const uint8_t SM2135_30MA      =  0x04;  // CW max current (Chip default)
+static const uint8_t SM2135_35MA      =  0x05;
+static const uint8_t SM2135_40MA      =  0x06;
+static const uint8_t SM2135_45MA      =  0x07;  // Max value for RGB
+static const uint8_t SM2135_50MA      =  0x08;
+static const uint8_t SM2135_55MA      =  0x09;
+static const uint8_t SM2135_60MA      =  0x0A;
+
+static const uint8_t SM2135_CURRENT = (SM2135_20MA << 4) | SM2135_10MA;
+
+void SM2135::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up SM2135OutputComponent...");
+  this->data_pin_->setup();
+  this->data_pin_->digital_write(true);
+  this->clock_pin_->setup();
+  this->clock_pin_->digital_write(true);
+  this->pwm_amounts_.resize(5, 0);
+}
+void SM2135::dump_config() {
+  ESP_LOGCONFIG(TAG, "SM2135:");
+  LOG_PIN("  Data Pin: ", this->data_pin_);
+  LOG_PIN("  Clock Pin: ", this->clock_pin_);
+  ESP_LOGCONFIG(TAG, "  Total number of channels: 5");
+}
+
+void SM2135::loop() {
+  if (!this->update_)
+    return;
+
+  uint8_t data[6];
+  if (this->update_channel_ == 4 || this->update_channel_ == 5) {
+    // No color so must be Cold/Warm
+    data[0] = SM2135_ADDR_MC;
+    data[1] = SM2135_CURRENT;
+    data[2] = SM2135_CW;
+    this->write_buffer_(data, 3);
+    delay(1);
+    data[0] = SM2135_ADDR_C;
+    data[1] = this->pwm_amounts_[4];  // Warm
+    data[2] = this->pwm_amounts_[3];  // Cold
+    this->write_buffer_(data, 3);
+  } else {
+    // Color
+    data[0] = SM2135_ADDR_MC;
+    data[1] = SM2135_CURRENT;
+    data[2] = SM2135_RGB;
+    data[3] = this->pwm_amounts_[1];  // Green
+    data[4] = this->pwm_amounts_[0];  // Red
+    data[5] = this->pwm_amounts_[2];  // Blue
+    this->write_buffer_(data, 6);
+  }
+
+  this->update_ = false;
+}
+
+}  // namespace sm2135
+}  // namespace esphome

--- a/esphome/components/sm2135/sm2135.cpp
+++ b/esphome/components/sm2135/sm2135.cpp
@@ -53,7 +53,7 @@ void SM2135::loop() {
     return;
 
   uint8_t data[6];
-  if (this->update_channel_ == 4 || this->update_channel_ == 5) {
+  if (this->update_channel_ == 3 || this->update_channel_ == 4) {
     // No color so must be Cold/Warm
     data[0] = SM2135_ADDR_MC;
     data[1] = SM2135_CURRENT;

--- a/esphome/components/sm2135/sm2135.cpp
+++ b/esphome/components/sm2135/sm2135.cpp
@@ -1,37 +1,35 @@
 #include "sm2135.h"
 #include "esphome/core/log.h"
 
-
 // Tnx to the work of https://github.com/arendst (Tasmota) for making the initial version of the driver
-
 
 namespace esphome {
 namespace sm2135 {
 
 static const char *TAG = "sm2135";
 
-static const uint8_t SM2135_ADDR_MC   =  0xC0;  // Max current register
-static const uint8_t SM2135_ADDR_CH   =  0xC1;  // RGB or CW channel select register
-static const uint8_t SM2135_ADDR_R    =  0xC2;  // Red color
-static const uint8_t SM2135_ADDR_G    =  0xC3;  // Green color
-static const uint8_t SM2135_ADDR_B    =  0xC4;  // Blue color
-static const uint8_t SM2135_ADDR_C    =  0xC5;  // Cold
-static const uint8_t SM2135_ADDR_W    =  0xC6;  // Warm
+static const uint8_t SM2135_ADDR_MC = 0xC0;  // Max current register
+static const uint8_t SM2135_ADDR_CH = 0xC1;  // RGB or CW channel select register
+static const uint8_t SM2135_ADDR_R = 0xC2;   // Red color
+static const uint8_t SM2135_ADDR_G = 0xC3;   // Green color
+static const uint8_t SM2135_ADDR_B = 0xC4;   // Blue color
+static const uint8_t SM2135_ADDR_C = 0xC5;   // Cold
+static const uint8_t SM2135_ADDR_W = 0xC6;   // Warm
 
-static const uint8_t SM2135_RGB       =  0x00;  // RGB channel
-static const uint8_t SM2135_CW        =  0x80;  // CW channel (Chip default)
+static const uint8_t SM2135_RGB = 0x00;  // RGB channel
+static const uint8_t SM2135_CW = 0x80;   // CW channel (Chip default)
 
-static const uint8_t SM2135_10MA      =  0x00;
-static const uint8_t SM2135_15MA      =  0x01;
-static const uint8_t SM2135_20MA      =  0x02;  // RGB max current (Chip default)
-static const uint8_t SM2135_25MA      =  0x03;
-static const uint8_t SM2135_30MA      =  0x04;  // CW max current (Chip default)
-static const uint8_t SM2135_35MA      =  0x05;
-static const uint8_t SM2135_40MA      =  0x06;
-static const uint8_t SM2135_45MA      =  0x07;  // Max value for RGB
-static const uint8_t SM2135_50MA      =  0x08;
-static const uint8_t SM2135_55MA      =  0x09;
-static const uint8_t SM2135_60MA      =  0x0A;
+static const uint8_t SM2135_10MA = 0x00;
+static const uint8_t SM2135_15MA = 0x01;
+static const uint8_t SM2135_20MA = 0x02;  // RGB max current (Chip default)
+static const uint8_t SM2135_25MA = 0x03;
+static const uint8_t SM2135_30MA = 0x04;  // CW max current (Chip default)
+static const uint8_t SM2135_35MA = 0x05;
+static const uint8_t SM2135_40MA = 0x06;
+static const uint8_t SM2135_45MA = 0x07;  // Max value for RGB
+static const uint8_t SM2135_50MA = 0x08;
+static const uint8_t SM2135_55MA = 0x09;
+static const uint8_t SM2135_60MA = 0x0A;
 
 static const uint8_t SM2135_CURRENT = (SM2135_20MA << 4) | SM2135_10MA;
 

--- a/esphome/components/sm2135/sm2135.cpp
+++ b/esphome/components/sm2135/sm2135.cpp
@@ -45,7 +45,6 @@ void SM2135::dump_config() {
   ESP_LOGCONFIG(TAG, "SM2135:");
   LOG_PIN("  Data Pin: ", this->data_pin_);
   LOG_PIN("  Clock Pin: ", this->clock_pin_);
-  ESP_LOGCONFIG(TAG, "  Total number of channels: 5");
 }
 
 void SM2135::loop() {

--- a/esphome/components/sm2135/sm2135.h
+++ b/esphome/components/sm2135/sm2135.h
@@ -1,0 +1,87 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/core/esphal.h"
+#include "esphome/components/output/float_output.h"
+
+namespace esphome {
+namespace sm2135 {
+
+class SM2135 : public Component {
+ public:
+  class Channel;
+
+  void set_data_pin(GPIOPin *data_pin) { data_pin_ = data_pin; }
+  void set_clock_pin(GPIOPin *clock_pin) { clock_pin_ = clock_pin; }
+  void set_num_channels(uint8_t num_channels) { num_channels_ = num_channels; }
+  void set_num_chips(uint8_t num_chips) { num_chips_ = num_chips; }
+
+  void setup() override;
+
+  void dump_config() override;
+
+  float get_setup_priority() const override { return setup_priority::HARDWARE; }
+
+  /// Send new values if they were updated.
+  void loop() override;
+
+  class Channel : public output::FloatOutput {
+   public:
+    void set_parent(SM2135 *parent) { parent_ = parent; }
+    void set_channel(uint8_t channel) { channel_ = channel; }
+
+   protected:
+    void write_state(float state) override {
+        auto amount = static_cast<uint8_t>(state * 0xff);
+        this->parent_->set_channel_value_(this->channel_, amount);
+    }
+
+    SM2135 *parent_;
+    uint8_t channel_;
+  };
+
+ protected:
+  void set_channel_value_(uint8_t channel, uint8_t value) {
+    if (this->pwm_amounts_[channel] != value) {
+      this->update_ = true;
+      this->update_channel_ = channel;
+    }
+    this->pwm_amounts_[channel] = value;
+  }
+  void write_bit_(bool value) {
+    this->clock_pin_->digital_write(false);
+    this->data_pin_->digital_write(value);
+    this->clock_pin_->digital_write(true);
+  }
+
+  void write_byte_(uint8_t data) {
+    for (uint8_t mask = 0x80; mask; mask >>= 1) {
+        this->write_bit_(data & mask);
+    }
+    this->clock_pin_->digital_write(false);
+    this->data_pin_->digital_write(true);
+    this->clock_pin_->digital_write(true);
+  }
+
+  void write_buffer_(uint8_t *buffer, uint8_t size) {
+    this->data_pin_->digital_write(false);
+    for (uint32_t i = 0; i < size; i++) {
+        this->write_byte_(buffer[i]);
+    }
+    this->clock_pin_->digital_write(false);
+    this->clock_pin_->digital_write(true);
+    this->data_pin_->digital_write(true);
+    
+  }
+
+  GPIOPin *data_pin_;
+  GPIOPin *clock_pin_;
+  uint8_t num_channels_;
+  uint8_t num_chips_;
+  uint8_t update_channel_;
+  std::vector<uint8_t> pwm_amounts_;
+  bool update_{true};
+};
+
+}  // namespace sm2135
+}  // namespace esphome

--- a/esphome/components/sm2135/sm2135.h
+++ b/esphome/components/sm2135/sm2135.h
@@ -13,7 +13,6 @@ class SM2135 : public Component {
 
   void set_data_pin(GPIOPin *data_pin) { data_pin_ = data_pin; }
   void set_clock_pin(GPIOPin *clock_pin) { clock_pin_ = clock_pin; }
-  void set_num_channels(uint8_t num_channels) { num_channels_ = num_channels; }
 
   void setup() override;
 
@@ -74,7 +73,6 @@ class SM2135 : public Component {
 
   GPIOPin *data_pin_;
   GPIOPin *clock_pin_;
-  uint8_t num_channels_;
   uint8_t update_channel_;
   std::vector<uint8_t> pwm_amounts_;
   bool update_{true};

--- a/esphome/components/sm2135/sm2135.h
+++ b/esphome/components/sm2135/sm2135.h
@@ -32,8 +32,8 @@ class SM2135 : public Component {
 
    protected:
     void write_state(float state) override {
-        auto amount = static_cast<uint8_t>(state * 0xff);
-        this->parent_->set_channel_value_(this->channel_, amount);
+      auto amount = static_cast<uint8_t>(state * 0xff);
+      this->parent_->set_channel_value_(this->channel_, amount);
     }
 
     SM2135 *parent_;
@@ -56,7 +56,7 @@ class SM2135 : public Component {
 
   void write_byte_(uint8_t data) {
     for (uint8_t mask = 0x80; mask; mask >>= 1) {
-        this->write_bit_(data & mask);
+      this->write_bit_(data & mask);
     }
     this->clock_pin_->digital_write(false);
     this->data_pin_->digital_write(true);
@@ -66,12 +66,11 @@ class SM2135 : public Component {
   void write_buffer_(uint8_t *buffer, uint8_t size) {
     this->data_pin_->digital_write(false);
     for (uint32_t i = 0; i < size; i++) {
-        this->write_byte_(buffer[i]);
+      this->write_byte_(buffer[i]);
     }
     this->clock_pin_->digital_write(false);
     this->clock_pin_->digital_write(true);
     this->data_pin_->digital_write(true);
-    
   }
 
   GPIOPin *data_pin_;

--- a/esphome/components/sm2135/sm2135.h
+++ b/esphome/components/sm2135/sm2135.h
@@ -14,7 +14,6 @@ class SM2135 : public Component {
   void set_data_pin(GPIOPin *data_pin) { data_pin_ = data_pin; }
   void set_clock_pin(GPIOPin *clock_pin) { clock_pin_ = clock_pin; }
   void set_num_channels(uint8_t num_channels) { num_channels_ = num_channels; }
-  void set_num_chips(uint8_t num_chips) { num_chips_ = num_chips; }
 
   void setup() override;
 
@@ -76,7 +75,6 @@ class SM2135 : public Component {
   GPIOPin *data_pin_;
   GPIOPin *clock_pin_;
   uint8_t num_channels_;
-  uint8_t num_chips_;
   uint8_t update_channel_;
   std::vector<uint8_t> pwm_amounts_;
   bool update_{true};

--- a/tests/test3.yaml
+++ b/tests/test3.yaml
@@ -586,6 +586,10 @@ script:
     then:
       - lambda: 'ESP_LOGD("main", "Hello World!");'
 
+sm2135:
+  data_pin: GPIO12
+  clock_pin: GPIO14
+
 switch:
   - platform: template
     name: 'mpr121_toggle'

--- a/tests/test3.yaml
+++ b/tests/test3.yaml
@@ -832,6 +832,21 @@ output:
     pin: GPIO5
     id: my_slow_pwm
     period: 15s
+  - platform: sm2135
+    id: sm2135_0
+    channel: 0
+  - platform: sm2135
+    id: sm2135_1
+    channel: 1
+  - platform: sm2135
+    id: sm2135_2
+    channel: 2
+  - platform: sm2135
+    id: sm2135_3
+    channel: 3
+  - platform: sm2135
+    id: sm2135_4
+    channel: 4
 
 mcp23017:
   id: mcp23017_hub

